addon/devtoolsPanel.js

```diff
@@ -1,47 +1,232 @@
-    backgroundPort.postMessage({type: 'label',
-                                tabId: browser.devtools.inspectedWindow.tabId,
-                                inspectedElement: await inspectedElementPath(),
-                                label: document.getElementById('labelField').value});
+// Returns a list of objects with the following attributes:
+//   preview: html preview of element
+//   label: the fathom label, or an empty string if unlabeled
+//   path: css path to element
+//   inspected: true if this element === $0
+function getLabeled() {
+    let evalScript = `
+        (function getLabeled() {
+            function elementMeta(el, isInspected) {
+                return {
+                    preview: el.outerHTML.replace(/^([^>]+>)[\\s\\S]*$/, '$1'),
+                    label: el.dataset.fathom || '',
+                    path: Simmer(el),
+                    inspected: isInspected,
+                };
+            }
+
+            let list = [];
+            if ($0) {
+                list.push(elementMeta($0, true));
+            }
+            for (const el of document.querySelectorAll('[data-fathom]')) {
+                if ($0 && el === $0) {
+                    continue;
+                }
+                list.push(elementMeta(el, false));
+            }
+            return list;
+        })();
+    `;
+    return resultOfEval(evalScript);
 }
-document.getElementById('labelButton').addEventListener('click', labelInspectedElement);
 
-async function freezePage() {
-    backgroundPort.postMessage({type: 'freeze',
-                                tabId: browser.devtools.inspectedWindow.tabId,
-                                inspectedElement: await inspectedElementPath(),
-                                options: {shouldScroll: false, wait: 0}});
+// Draw the table with the list of existing labeled elements.
+// If an element is inspected/selected with devtools (i.e. $0 is defined), it
+// will always be the first row in the table.
+function updateLabeledTable(labeled) {
+
+    function addTextCell(row, text, isHeader=false) {
+        const td = document.createElement(isHeader ? 'th' : 'td');
+        td.appendChild(document.createTextNode(text));
+        row.appendChild(td);
+        return td;
+    }
+
+    function inspectLabeled(event) {
+        event.preventDefault();
+        const js = 'inspect(document.querySelector("' + event.target.dataset.path + '"))';
+        browser.devtools.inspectedWindow.eval(js)
+            .catch((e) => {
+                console.error(e);
+            });
+    }
+
+    function addPathCell(row, label) {
+        const td = document.createElement('td');
+        td.classList.add('path');
+        if (label.inspected) {
+            td.appendChild(document.createTextNode(label.path));
+        } else {
+            let a = document.createElement('a');
+            a.href = '#';
+            a.dataset.path = label.path;
+            a.addEventListener('click', inspectLabeled);
```

Comment:
Wow, this is all really crying out for some terser and/or more declarative framework. You have it about as tight as you can get it procedurally, I think. A future thing to think about.